utils/mutex_map.go

```diff
@@ -0,0 +1,129 @@
+package utils
+
+import (
+	"sync"
+	"sync/atomic"
+)
+
+// MutexMap provides a locking mechanism based on provided name
+type MutexMap struct {
+	mutex sync.Mutex
+	locks map[string]*lockEntry
+}
+
+// NewMutexMap creates a new MutexMap
+func NewMutexMap() *MutexMap {
+	return &MutexMap{
+		locks: make(map[string]*lockEntry),
+	}
+}
+
+func (m *MutexMap) lock(name string, readonly bool) {
+	m.mutex.Lock()
+	nameLock, exists := m.locks[name]
+	if !exists {
+		nameLock = &lockEntry{
+			waiters: 1,
+		}
+		m.locks[name] = nameLock
+	} else {
+		nameLock.inc()
+	}
+
+	m.mutex.Unlock()
+	if readonly {
+		nameLock.rlock()
+	} else {
+		nameLock.lock()
+	}
+	// I used the resource, I am not waiting any more to access the lock
+	nameLock.dec()
+}
+
+func (m *MutexMap) RLock(name string) {
+	m.lock(name, true)
+}
+
+// Lock locks a mutex with the given name.
+func (m *MutexMap) Lock(name string) {
+	m.lock(name, false)
+}
+
+func (m *MutexMap) LockAll() {
+	m.mutex.Lock()
```

Comment:
This does not really lock all, it just blocks acquiring new locks, but old locks will still be held and other threads can be reading/updating the cache keys.

Reply:
It should work.
There are two locks: one for map access and every map entry has its own lock (lock per cache key)

```
type MutexMap struct {
	mutex sync.Mutex (map access lock)
	locks map[string]*lockEntry (lock per cache key)
}
```

Every lock (or unlock) per map key requires first to gather map lock.

```
func (m *MutexMap) lock(name string, readonly bool) {
	m.mutex.Lock()
```
```
func (m *MutexMap) unlock(name string, readonly bool) {
	m.mutex.Lock()	
```
So if `LockAl`l is called other threads cannot access key locks in map.  I only need this method to purge the cache. To lock everything when redis cache is purged